src/psyclone/domain/lfric/metadata_to_arguments_rules.py

```diff
@@ -0,0 +1,614 @@
+# BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+# LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+# LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+# -----------------------------------------------------------------------------
+# Author: R. W. Ford, STFC Daresbury Lab
+
+'''This module implements a class that encapsulates rules that map
+   LFRic kernel metadata to kernel arguments.
+
+'''
+from psyclone.domain.lfric import LFRicConstants
+from psyclone.domain.lfric.kernel import (
+    OperatorArgMetadata, ColumnwiseOperatorArgMetadata, FieldArgMetadata,
+    FieldVectorArgMetadata, InterGridArgMetadata, InterGridVectorArgMetadata,
+    ScalarArgMetadata)
+from psyclone.errors import InternalError
+
+# pylint: disable=too-many-branches
+# pylint: disable=too-many-statements
+# pylint: disable=too-few-public-methods
+
+
+class MetadataToArgumentsRules():
+    '''This class encapsulates rules to map LFRic kernel metadata to
+    kernel arguments. It does this by calling class methods each of
+    which which represent a particular kernel argument or set of
+    arguments. It calls these in the order that the arguments should
+    be found in the kernel metadata. The particular methods called and
+    order that these methods are called are determined by the supplied
+    kernel metadata.
+
+    Kernel argument information from kernel metadata can be used for
+    more than one thing, e.g. to create or check arguments within a
+    kernel and their declarations (using PSyIR) or to create the
+    arguments in the calling method from the generated PSy-layer code,
+    which at some point will (mostly) need to be dereferenced from
+    existing datatypes, such as fields, or to create appropriate
+    algorithm PSyIR for a) the calling routine or b) the subroutine
+    body in the PSy-layer. Subclasses of this class can be implemented
+    for these different requirements.
+
+    '''
+    _metadata = None
+    _info = None
+
+    @classmethod
+    def mapping(cls, metadata, info=None):
+        '''Takes kernel metadata as input and returns whatever is added to the
+        _info variable. This class adds nothing to the _info variable,
+        it is up to the subclass to do this. The variable is treated
+        as a container. The optional info argument allows the subclass
+        to add to an existing object if required.
+
+        :param metadata: the kernel metadata.
+        :type metadata: \
+            py:class:`psyclone.domain.lfric.kernel.LFRicKernelMetadata`
+        :param info: optional object to initialise the _info \
+            variable. Defaults to None.
+        :type info: :py:class:`Object`
+
+        '''
+        cls._initialise(info)
+        cls._metadata = metadata
```

Comment:
I think I was aiming for a fully stateless internal solution then backtracked. I've changed it so that metadata is not passed into `_generate()`. by argument. The reason to keep a reference to the metadata is that it might be needed by a subclass method. In theory the metadata could be passed by argument to all of the methods just in case, but that seems wrong to me.